pkg/ingester/client/pool_test.go

```diff
@@ -0,0 +1,125 @@
+package client
+
+import (
+	fmt "fmt"
+	"testing"
+	"time"
+
+	context "golang.org/x/net/context"
+	grpc "google.golang.org/grpc"
+	"google.golang.org/grpc/health/grpc_health_v1"
+)
+
+type mockIngester struct {
+	IngesterClient
+	happy  bool
+	status grpc_health_v1.HealthCheckResponse_ServingStatus
+}
+
+func (i mockIngester) Check(ctx context.Context, in *grpc_health_v1.HealthCheckRequest, opts ...grpc.CallOption) (*grpc_health_v1.HealthCheckResponse, error) {
+	if !i.happy {
+		return nil, fmt.Errorf("Fail")
+	}
+	return &grpc_health_v1.HealthCheckResponse{Status: i.status}, nil
+}
+
+func (i mockIngester) Close() error {
+	return nil
+}
+
+func TestHealthCheck(t *testing.T) {
+	tcs := []struct {
+		ingester mockIngester
+		hasError bool
+	}{
+		{mockIngester{happy: true, status: grpc_health_v1.HealthCheckResponse_UNKNOWN}, true},
+		{mockIngester{happy: true, status: grpc_health_v1.HealthCheckResponse_SERVING}, false},
+		{mockIngester{happy: true, status: grpc_health_v1.HealthCheckResponse_NOT_SERVING}, true},
+		{mockIngester{happy: false, status: grpc_health_v1.HealthCheckResponse_UNKNOWN}, true},
+		{mockIngester{happy: false, status: grpc_health_v1.HealthCheckResponse_SERVING}, true},
+		{mockIngester{happy: false, status: grpc_health_v1.HealthCheckResponse_NOT_SERVING}, true},
+	}
+	for _, tc := range tcs {
+		err := healthCheck(tc.ingester, 50*time.Millisecond)
+		hasError := err != nil
+		if hasError != tc.hasError {
+			t.Errorf("Expected error: %t, error: %v", tc.hasError, err)
+		}
+	}
+}
+
+func TestIngesterCache(t *testing.T) {
+	buildCount := 0
+	factory := func(addr string, _ Config) (IngesterClient, error) {
+		if addr == "bad" {
+			return nil, fmt.Errorf("Fail")
+		}
+		buildCount++
+		return mockIngester{happy: true, status: grpc_health_v1.HealthCheckResponse_SERVING}, nil
+	}
+	pool := NewIngesterPool(factory, Config{}, 50*time.Millisecond)
+
+	pool.GetClientFor("1")
+	if buildCount != 1 {
+		t.Errorf("Did not create client")
+	}
+
+	pool.GetClientFor("1")
+	if buildCount != 1 {
+		t.Errorf("Created client that should have been cached")
+	}
+
+	pool.GetClientFor("2")
+	if pool.Count() != 2 {
+		t.Errorf("Expected Count() = 2, got %d", pool.Count())
```

Comment:
This is a bit wordy - we use `testify/assert` elsewhere to reduce to `assert.Equal(t, 2, pool.Count())`